orchestra/cmds/components.py

```diff
@@ -70,21 +74,79 @@ def handle_components(args):
+        installed_build = metadata and metadata.get("build_name")
+
+        if args.installed and installed_build is None \
+                or args.not_installed and installed_build is not None:
+            continue
+
+        components_to_print.add(component)
+
+    if args.json:
+        print_json(components_to_print, config)
+    else:
+        print_human_readable(components_to_print, config, args)
+
+
+def print_json(components, config):
+    components_json = []
+    for component in components:
+        component_name = component.name
         metadata = get_installed_metadata(component_name, config)
         is_installed = metadata is not None
         manually_installed = is_installed and metadata.get("manually_installed", False)
         installed_build = metadata and metadata.get("build_name")
-        if args.installed and not is_installed:
-            continue
-        if args.not_installed and is_installed:
-            continue
+        component_info = {
+            "name": component.name,
+            "installed": is_installed,
+            "manually_installed": manually_installed,
+            "installed_build_name": installed_build,
+            "hash": component.self_hash,
+            "recursive_hash": component.recursive_hash,
+            "default_build": component.default_build.name,
+            "builds": {},
+        }
+
+        if component.clone:
+            branch_name, branch_commit = component.clone.branch()
+            component_info["head_branch_name"] = branch_name
+            component_info["head_commit"] = branch_commit
+
+        for build_name, build in component.builds.items():
+            build_info = {
+                "installed": installed_build == build_name,
+                "default": build is component.default_build,
+                "qualified_name": build.qualified_name,
+            }
+            if build.configure:
+                build_info["config_deps"] = [d.name_for_components for d in build.configure.dependencies]
+
+            if build.install:
+                build_info["install_deps"] = [d.name_for_components for d in build.install.dependencies]
+
+            component_info["builds"][build_name] = build_info
+        components_json.append(component_info)
+    print(json.dumps(components_json))
+
+
+def print_human_readable(components, config, args):
+    for component in components:
+        component_name = component.name
+
+        installed_build = get_installed_build(component_name, config)
 
         component_infos = []
         if args.hashes:
             component_infos.append(f"hash: {component.self_hash}")
             component_infos.append(f"recursive hash: {component.recursive_hash}")
         component_infos_s = stringify_infos(component_infos)
 
+        metadata = get_installed_metadata(component_name, config)
+        is_installed = metadata is not None
+        manually_installed = is_installed and metadata.get("manually_installed", False)
```

Comment:
Duplicated code

Reply:
What do you suggest? Creating a function that returns a tuple does not look much better to me honestly
```
def get_install_type(component_name, config):
    metadata = get_installed_metadata(component_name, config)
    is_installed = metadata is not None
    manually_installed = is_installed and metadata.get("manually_installed", False)
    return is_installed, manually_installed
```

A function returning the installation type is also more verbose (and it gets worse if we want to use enums and not strings, which I personally dislike):
```
def install_type(component_name, config):
    metadata = get_installed_metadata(component_name, config)
    is_installed = metadata is not None
    manually_installed = is_installed and metadata.get("manually_installed", False)
    if not is_installed:
        return None
    elif manually_installed:
        return "manually"
    elif:
        return "dependency"
```

Having two functions `is_installed` and `manually_installed` seems a bit wasteful, needing to parse the JSON twice (or cache it, increasing complexity). I'm open to suggestions :)

Reply:
The first option looks OK to me.

Reply:
I did not see it before commenting, `print_json` also uses `metadata` just below those three lines, the first approach would become a Frankenstein function which returns metadata and some data derived from it, it's not good.
I can go for the second, re-reading medatata every time, or create functions which take metadata and extract the properties from it, seems overkill to me for avoiding a two-line repetition, WDYT?

Reply:
At this point I guess a better solution is to create a Metadata class, I'll do that

Reply:
I created the InstallMetadata class which handles the metadata about installed components, waiting for review